### DISCUSSION TITLE
Full file URL in API

### DISCUSSION THREAD

**Discussion by digitaljohn** [directus/directus#5928] [Feature Requests]
Currently, when trying to get information on files/images using the GraphQL API it only returns fields like the asset ID. This means we need another mechanism to retrieve the full URL for the asset and prepend the absolute path of the Directus server.

Is it possible to add a new field in the GraphQL schema like `url` which contains the absolute path by utilising the `env.PUBLIC_URL`?

Thoughts?

**Comment by benhaynes** [comment]
You're looking for this: https://docs.directus.io/reference/api/query/#fields

**Reply by digitaljohn** [reply]
@benhaynes maybe I am missing something? In the screenshot below I'm outputting every field for a file but cannot see anything that would include the full absolute image URL?

![image](https://user-images.githubusercontent.com/1425921/120129034-2d6b8b80-c1bb-11eb-8919-5b2087ffa1ce.png)

**Reply by benhaynes** [reply]
While you _can_ use the actual asset URL from the `filename_disk` and your configured asset URL, the `id` can be used to fetch it through the API:

https://docs.directus.io/reference/api/assets/#accessing-an-original-file

**Reply by digitaljohn** [reply]
Understood @benhaynes, I just feel the API should be able to return a full asset URL without needing the frontend to manually construct it as the CMS has full knowledge to be able to do it when `PUBLIC_URL` is specified.

Moreover, I think this should be made available as it seems the Directus team is heavily recommending loading assets directly from Directus. https://github.com/directus/directus/discussions/5529#discussioncomment-709560

**Reply by philleepflorence** [reply]
@digitaljohn ... Please see if the answer help...

**Reply by benhaynes** [reply]
> Understood @benhaynes, I just feel the API should be able to return a full asset URL without needing the frontend to manually construct it as the CMS has full knowledge to be able to do it when `PUBLIC_URL` is specified. Moreover, I think this should be made available as it seems the Directus team is heavily recommending loading assets directly from Directus. [#5529 (reply in thread)](https://github.com/directus/directus/discussions/5529#discussioncomment-709560)

I think this is useful information to return, perhaps as nested data within the `storage` object/key... or as an alias field. @rijkvanzanten ?

**Reply by rijkvanzanten** [reply]
`storage` is the used storage adapter (key), so we won't be nesting things under there. We can look into adding a "presentation field" to directus_files that uses a custom transform to stitch together the file id with the public url path to assets

**Reply by benhaynes** [reply]
Yeah, I think that presentation field is the best option.

**Comment by philleepflorence** [comment]
#### Background on Files and Assets

> [Files](https://docs.directus.io/reference/api/system/files/) are the objects that contain information about the file managed through Directus. 
> The `/assets` [endpoint](https://docs.directus.io/reference/api/assets/) can be used to stream or retrieve the actual file contents from assets managed within Directus. 
```
example.com/assets/<file-id>
example.com/assets/1ac73658-8b62-4dea-b6da-529fbc9d01a4

example.com/assets/<file-id>?key=<preset-key>
example.com/assets/1ac73658-8b62-4dea-b6da-529fbc9d01a4?key=card

example.com/assets/<file-id>?fit=<fit>&width=<width>&height=<height>&quality=<quality>
example.com/assets/1ac73658-8b62-4dea-b6da-529fbc9d01a4?fit=cover&width=200&height=200&quality=80
```
---

#### Existing Discussions
[#3182  Add full resolved files URL](https://github.com/directus/directus/discussions/3182)
[#5808 File & related collection references in API](https://github.com/directus/directus/discussions/5808)
[#4622 How do I use an uploaded file?](https://github.com/directus/directus/discussions/4622)
[#2883 Access full_url of the uploaded images](https://github.com/directus/directus/discussions/2883)

---

#### Options for Current Implementation  
- HTML Meta Elements - [Document Base URL](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base).
- HTML Embed Elements - [Picture Element Sources](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture).

#### Options for Similar Implementation
- API Response [Metadata](https://docs.directus.io/reference/api/query/#metadata) - links, formats, or options for  "documentation purposes".
- [Directus JavaScript SDK](https://docs.directus.io/reference/sdk/) - Compile URL or String.

---

#### User and Value Context  
_What different users expected of the same thing._
<details>
<summary>Read More...</summary>
- Some content creators _expected_ `slug interface` was also for Social or SEO URL fragments.  
- Some administrators _expected_ Project Settings or Collections Settings to manage URLs added in runtime by API Endpoint or Extension.
- Some users complained about removing fields or related; to add optional or related levels, not as much. 
</details>

#### Related Implementation
_How some users suggested the story._
<details>
<summary>Read More...</summary>
- For links managed by Directus or stored in the DB;
    - Use the `slug` options in [Field Input Interface](https://github.com/directus/directus/tree/main/app/src/interfaces/input)...
- For links that varied by Display Experience or Website;
    - Use [Directus SDK](https://docs.directus.io/reference/sdk/)
    - Use [Directus API Utils](https://github.com/directus/directus/tree/main/app/src/utils)
</details>

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
